pghoard/transfer.py

```diff
@@ -4,29 +4,105 @@
 Copyright (c) 2016 Ohmu Ltd
 See LICENSE for details
 """
+import dataclasses
+import enum
 import logging
 import os
 import time
+from dataclasses import dataclass
+from io import BytesIO
+from pathlib import Path
 from queue import Empty
 from threading import Lock, Thread
+from typing import Any, BinaryIO, Dict, Optional, Union
 
-from pghoard.common import create_alert_file, get_object_storage_config
+from pghoard.common import (
+    CallbackEvent, CallbackQueue, FileType, Queue, QuitEvent, StrEnum, create_alert_file, get_object_storage_config
+)
 from pghoard.fetcher import FileFetchManager
 from pghoard.rohmu import get_transfer
 from pghoard.rohmu.compat import suppress
-from pghoard.rohmu.errors import (FileNotFoundFromStorageError, LocalFileIsRemoteFileError)
+from pghoard.rohmu.errors import FileNotFoundFromStorageError
 
 _STATS_LOCK = Lock()
 _last_stats_transmit_time = 0
 
 
+@enum.unique
+class TransferOperation(StrEnum):
+    Download = "download"
+    Upload = "upload"
+    List = "list"
+    Metadata = "metadata"
+
+
+# Setup a really complicated class hierarchy to handle dataclass mixin
+# of default and non default fields...
+@dataclass(frozen=True)
+class BaseTransferEvent:
+    backup_site_key: str
+    file_type: FileType
+    file_path: Path
+    callback_queue: CallbackQueue
+
+
+@dataclass(frozen=True)
+class UploadEvent(BaseTransferEvent):
+    source_data: Union[BinaryIO, Path]
+    metadata: Dict[str, str]
+    file_size: Optional[int]
+    remove_after_upload: bool = True
+    retry_number: int = 0
+
+    @property
+    def operation(self):
+        return TransferOperation.Upload
+
+
+@dataclass(frozen=True)
+class DownloadEvent(BaseTransferEvent):
+
+    destination_path: Path
+    opaque: Optional[Any] = None
+    file_size: Optional[int] = 0
+
+    @property
+    def operation(self):
+        return TransferOperation.Download
+
+
+@dataclass(frozen=True)
+class ListEvent(BaseTransferEvent):
+    @property
+    def operation(self):
+        return TransferOperation.List
+
+
+@dataclass(frozen=True)
+class MetadataEvent(BaseTransferEvent):
+    @property
+    def operation(self):
+        return TransferOperation.Metadata
+
+
+OperationEvents = {
+    TransferOperation.Download: DownloadEvent,
+    TransferOperation.Upload: UploadEvent,
+    TransferOperation.List: ListEvent,
+    TransferOperation.Metadata: MetadataEvent
+}
+
+# Should be changed to Queue[TransferEvent] once
```

Comment:
nit. Since `QuitEvent` is separate, comment should say `Queue[Union[TransferEvent, Literal[QuitEvent]]]`